packages/rxsocketio/src/operators/send.js

```diff
@@ -11,24 +31,75 @@ function getMessageArgs(message) {
   return [topic, formattedMessage, ...binary];
 }
 
+function sendWithTransactions({
+  message,
+  socket,
+  action,
+  // If true, messages will wait in line for the prior message to be received
+  // before the next is emitted. (It forces sequential data transfer.)
+  maxTimeout,
+  // Should take a response and return true if response is valid, otherwise
+  // false:
+  validateResponse,
+  shouldRetry,
+  retryDelay,
+  retryCount,
+  retryFunction, // custom function to pass to retryWhen operator
+}) {
+  const transaction$ = new Observable(obs => {
+    socket
+      .binary(!!message.binary)
+      .compress(!!message.binary)
+      .emit(...getMessageArgs(message), (...responseArgs) => {
+        if (!validateResponse(responseArgs)) {
+          return obs.error(new Error('Server did not acknowledge response'));
+        }
+        obs.next([socket, action]);
+        return obs.complete();
+      });
+  });
+  return transaction$.pipe(
+    timeout(maxTimeout),
+    (
+      shouldRetry
+      ? retryWhen(retryFunction || (
+          shouldRetry
+          ? error$ => error$.pipe(delay(retryDelay), take(retryCount))
+          : () => of()
+      ))
+      : tap(() => 1)
+    )
+  );
+}
+
+function sendMessage(params) {
+  const {action, message, shouldWaitForConfirmation, socket} = params;
+  if (shouldWaitForConfirmation) return sendWithTransactions(params);
+  socket.binary(!!message.binary)
+    .compress(!!message.binary)
+    .emit(...getMessageArgs(message));
+  return of([socket, action]);
+}
+
 const send = function send({
   io$, // stream of websocket events from the core ws() Observable
+  ...options
 }) {
+  if (!io$.subscribe) {
+    return throwError(new Error(
+      'send() operator takes an Observable as its first parameter.'
+    ));
+  }
+  const config = {...defaultOptions, ...options};
   // stream of data objects (any valid data type) to be sent to the server
   return messageIn$ => {
-    const wsEvent$ = (
-      io$.subscribe && io$.pipe
-      ? io$.pipe(shareReplay(1))
-      : throwError(new Error(
-        'send() operator takes an Observable as its first parameter.'
-      ))
-    );
+    const wsEvent$ = io$.pipe(shareReplay(1));
     const sendStream$ = messageIn$.pipe(
       withLatestFrom(wsEvent$),
-      mergeMap(([message, [socket, action]]) => {
-        socket.emit(...getMessageArgs(message));
-        return of([socket, action]);
-      })
+      map(([message, [socket, action]]) => (
+        sendMessage({message, socket, action, ...config})
+      )),
+      mergeAll(config.shouldWaitForConfirmation ? 1 : null)
```

Comment:
I borrowed your `mergeAll` trick here!  Now the message emitter can force messages to wait in line until the server confirms receipt of the prior message.

Reply:
In the case `config.shouldWaitFromConfirmation = false` does null work? It looks like the default value is `Number.POSITIVE_INFINITY`

Reply:
🕵️  Good question.  I'll doublecheck that.

Reply:
⚠️